File: LinkedList/LinkedList.java
/**
 * LinkedList implementation for data type <T>.
 * @author Yasin Radi
 */
public class LinkedList<T> {

    /**
     * Node class holding T data type implementation.
     * @param <T>   Type of the data
     */
    private class Node {
        // Data held by the node.
        public T data;

        // Pointer to the next node in the list.
        public Node next;

        // Node constructor
        public Node() {
            this.data = null;
            this.next = null;
        };

        public Node(T v) {
            this.data = v;
            this.next = null;
        };

        public Node(T v, Node next) {
            this.data = v;
            this.next = next;
        };
    };

    // Head node.
    private Node head;

    // Tail node.
    private Node tail;

    // Length list atrribute.
    private int length;

    public LinkedList() {
        this.length = 0;
        this.head = null;
        this.tail = null;
    };

    /**
     * Checks if the given index idx is in the list's bounds.
     * @param idx   int
     * @return      boolean
     */
    private boolean checkIndexBounds(int idx) {
        return (idx >= 0) && (idx < this.length);
    };

    /**
     * Adds a given node to the list.
     * @param node  Node
     */
    public void add(Node node) {
        if (this.head == null) {
            this.head = node;
            this.tail = node;
        } else {
            this.tail.next = node;
            this.tail = node;
        }

        this.length++;
    };

    /**
     * Adds a node to the list created
     * using a given <T> value.
     * @param v <T>
     */
    public void add(T v) {
        add(new Node(v));
    };

    /**
     * Displays the content of all nodes.
     */
    public void displayNodes() {
        Node node = this.head;
        StringBuilder sb = new StringBuilder();
        while (node != null) {
            sb.append(node.data).append(" ");
            node = node.next;
        };

        System.out.println(sb.toString());
    };

    /**
     * Gets the node before the given index.
     * @param idx   int
     * @return      Node
     */
    public Node getNodeBeforeIndex(int idx) {
        if(!checkIndexBounds(idx)) {
            return null;
        }

        Node curr = this.head;
        Node pre = new Node();
        for (int i = 0; i < idx; i++) {
            pre = curr;
            curr = curr.next;
        };

        return pre;
    };

    /**
     * Adds a node at the beginning of the list 
     * created from a given <T> value.
     * @param v <T>
     */
    public void addAtStart(T v) {
        if (this.head == null) {
            add(v);
            return;
        };

        Node node = new Node(v, this.head);
        this.head = node;
        this.length++;
    };

    /**
     * Adds a given Node at the beginning of the list.
     * @param node  Node
     */
    public void addAtStart(Node node) {
        if(this.head == null) {
            add(node);
            return;
        };

        node.next = this.head;
        this.head = node;
        this.length++;
    };

    /**
     * Adds a given Node at the given index idx.
     * @param idx   int
     * @param node  Node
     */
    public void addAtIndex(int idx, Node node) {
        if (!checkIndexBounds(idx)) {
            return;
        }

        Node pre = getNodeBeforeIndex(idx);
        Node curr = pre.next;
        pre.next = node;
        node.next = curr;
        this.length++;
    };

    /**
     * Adds a given Node created from <T> value v
     * at the given index idx.
     * @param idx   int
     * @param v     <T>
     */
    public void addAtIndex(int idx, T v) {
        addAtIndex(idx, new Node(v));
    };

    /**
     * Pops out the first element of the list.
     */
    public void popFirst() {
        this.head = this.head.next;
        this.length--;
    };

    /**
     * Pops out the last element of the list.
     */
    public void pop() {
        this.tail = getNodeBeforeIndex(this.length - 1);
        this.tail.next = null;
        this.length--;
    };

    /**
     * Pops out the Node at the given index idx.
     * @param idx   int
     */
    public void popFromIndex(int idx) {
        if (!checkIndexBounds(idx)) return;

        Node pre = getNodeBeforeIndex(idx);
        Node curr = pre.next;
        pre.next = curr.next;
        this.length--;
    };
};

File: LinkedList/LinkedList.cs
using System;
using System.Text;

/// <summary>
/// LinkedList implementation for data type < T >.
/// </summary>
public class LinkedList<T>
{
    /// <summary>
    /// Node class holding T data type implementation.
    /// </summary>
    /// <typeparam name="T">Type of data held by list.</typeparam>
    protected class Node
    {
        // Data held by node
        public T data;

        // Pointer to next node in list
        public Node next;

        // Constructor
        public Node() {
            this.data = default(T);
            this.next = null;
        }

        public Node(T v) {
            this.data = v;
            this.next = null;
        }

        public Node(T v, Node next) {
            this.data = v;
            this.next = next;
        }
    }

    // Head node.
    private Node head;

    // Tail node.
    private Node tail;

    // List's length attribute.
    private int length;

    public LinkedList() {
        this.length = 0;
        this.head = null;
        this.tail = null;
    }

    /// <summary>
    /// Checks if the given index is in list's bounds.
    /// </summary>
    /// <param name="idx">Index to check</param>
    /// <returns>True if is in list bounds, false otherwise.</returns>
    protected bool CheckIndexBounds(int idx) {
        return idx >= 0 && idx <= length;
    }

    /// <summary>
    /// Adds a given node to the list.
    /// </summary>
    /// <param name="node">Node to be added to list.</param>
    protected void Add(Node node) {
        if (head == null) {
            head = node;
            tail = node;
        } else {
            tail.next = node;
            tail = node;
        }

        length++;
    }

    /// <summary>
    /// Adds a node to the list given a value of type < T >.
    /// </summary>
    /// <param name="v">Value to be added to the list.</param>
    public void Add(T v) {
        Add(new Node(v));
    }

    /// <summary>
    /// Displays the content of all nodes.
    /// </summary>
    public void Display() {
        Node node = head;
        StringBuilder sb = new StringBuilder();
        while (node != null) {
            sb.Append(node.data).Append(" ");
            node = node.next;
        }

        Console.WriteLine(sb.ToString());
    }

    /// <summary>
    /// Gets the node before the given index.
    /// </summary>
    /// <param name="idx">Index consequtive to the desired node.</param>
    /// <returns>The node before the given index.</returns>
    protected Node GetNodeBeforeIndex(int idx) {
        if (!CheckIndexBounds(idx)) return null;

        Node curr = head;
        Node pre = new Node();
        for (int i = 0; i < idx; i++) {
            pre = curr;
            curr = curr.next;
        }

        return pre;
    }

    /// <summary>
    /// Adds a node at the beginning of the list
    /// created from a given < T > value.
    /// </summary>
    /// <param name="v">Value of the new node.</param>
    public void AddAtStart(T v) {
        if (head == null) {
            Add(v);
            return;
        }

        Node node = new Node(v, head);
        head = node;
        length++;
    }

    /// <summary>
    /// Adds a given Node< T > at the given index idx.
    /// </summary>
    /// <param name="idx">Index where the new node will be added.</param>
    /// <param name="node">Node to be added.</param>
    protected void AddAtIndex(int idx, Node node) {
        if (!CheckIndexBounds(idx)) return;

        Node pre = GetNodeBeforeIndex(idx);
        Node curr = pre.next;
        pre.next = node;
        node.next = curr;
        length++;
    }

    /// <summary>
    /// Adds a given value v at the given index idx.
    /// </summary>
    /// <param name="idx">Index where the new node will be added.</param>
    /// <param name="v">Value to be added.</param>
    public void AddAtIndex(int idx, T v) {
        AddAtIndex(idx, new Node(v));
    }

    /// <summary>
    /// Pops out the first element of the list.
    /// </summary>
    public void PopFirst() {
        head = head.next;
        length--;
    }

    /// <summary>
    /// Pops out the last element of the list.
    /// </summary>
    public void Pop() {
        tail = GetNodeBeforeIndex(length - 1);
        tail.next = null;
        length--;
    }

    /// <summary>
    /// Pops out the element at the given index idx.
    /// </summary>
    /// <param name="idx">Index of the element to be popped out.</param>
    public void PopFromIndex(int idx) {
        if (!CheckIndexBounds(idx)) return;

        Node pre = GetNodeBeforeIndex(idx);
        Node curr = pre.next;
        pre.next = curr.next;
        length--;
    }
}